pkg/controller/release/auth_provider.go

```diff
@@ -0,0 +1,245 @@
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	providerGCPAuthPlugin = "provider-gcp" // so that this is different than "gcp" that's already in client-go tree.
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin(providerGCPAuthPlugin, newProviderGCPAuthProvider); err != nil {
+		klog.Fatalf("Failed to register gcp auth plugin: %v", err)
+	}
+}
+
+var (
+	// defaultScopes:
+	// - cloud-platform is the base scope to authenticate to GCP.
+	// - userinfo.email is used to authenticate to GKE APIs with gserviceaccount
+	//   email instead of numeric uniqueID.
+	defaultScopes = []string{
+		"https://www.googleapis.com/auth/cloud-platform",
+		"https://www.googleapis.com/auth/userinfo.email"}
+)
+
+// gcpAuthProvider is an auth provider plugin that uses GCP credentials to provide
+// tokens for kubectl to authenticate itself to the apiserver. A sample json config
+// is provided below with all recognized options described.
+//
+// {
+//   'auth-provider': {
+//     # Required
+//     "name": "provider-gcp",
+//
+//     'config': {
+//       # Authentication options
+//       # These options are used while getting a token.
+//
+//       # comma-separated list of GCP API scopes. default value of this field
+//       # is "https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/userinfo.email".
+// 		 # to override the API scopes, specify this field explicitly.
+//       "scopes": "https://www.googleapis.com/auth/cloud-platform"
+//
+//       # Caching options
+//
+//       # Raw string data representing cached access token.
+//       "access-token": "ya29.CjWdA4GiBPTt",
+//       # RFC3339Nano expiration timestamp for cached access token.
+//       "expiry": "2016-10-31 22:31:9.123",
+//
+//       # golang reference time in the format that the expiration timestamp uses.
+//       # If omitted, defaults to time.RFC3339Nano
+//       "time-fmt": "2006-01-02 15:04:05.999999999"
+//     }
+//   }
+// }
+//
+type providerGCPAuthProvider struct {
+	tokenSource oauth2.TokenSource
+	persister   restclient.AuthProviderConfigPersister
+}
+
+func newProviderGCPAuthProvider(_ string, providerGCPConfig map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	ts, err := tokenSource(providerGCPConfig)
+	if err != nil {
+		return nil, err
+	}
+	cts, err := newCachedTokenSource(providerGCPConfig["access-token"], providerGCPConfig["expiry"], persister, ts, providerGCPConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &providerGCPAuthProvider{cts, persister}, nil
+}
+
+func tokenSource(providerGCPConfig map[string]string) (oauth2.TokenSource, error) {
+	// Google Application Credentials-based token source
+	scopes := parseScopes(providerGCPConfig)
+	ts, err := google.JWTAccessTokenSourceWithScope([]byte(providerGCPConfig["creds"]), scopes...)
+	if err != nil {
+		return nil, fmt.Errorf("cannot construct JWT token source: %v", err)
+	}
```

Comment:
Can we use `errors.Wrapf` instead of `fmt.Errorf`? We mostly use `github.com/pkg/errors` throughout the Crossplane codebases.